aiounifi/interfaces/api.py

```diff
@@ -22,66 +22,84 @@ class APIItems:
     obj_id_key: str
     path: str
     item_cls: Any
+    events: tuple = ()
+    process_messages: tuple = ()
+    remove_messages: tuple = ()
 
     def __init__(self, controller) -> None:
         """Initialize API items."""
         self.controller = controller
         self._items: dict[int | str, Any] = {}
         self._subscribers: list[SubscriptionType] = []
 
+        message_filter = self.process_messages + self.remove_messages
+        if self.events:
+            message_filter += (MessageKey.EVENT,)
+
+        controller.events.subscribe(
+            self.process_websocket_data,
+            message_filter,
+            event_filter=self.events or None,
+        )
+
     @final
     async def update(self) -> None:
         """Refresh data."""
         raw = await self.controller.request("get", self.path)
         self.process_raw(raw)
 
-    @final
-    def process_raw(self, raw: list[dict]) -> set:
-        """Process data."""
+    def process_raw(self, raw: list[dict[str, Any]]) -> set:
+        """Process full raw response."""
         new_items = set()
-
         for raw_item in raw:
-            key = raw_item[self.obj_id_key]
+            if obj_id := self.process_item(raw_item):
+                new_items.add(obj_id)
+        return new_items
 
-            if (obj := self._items.get(key)) is not None:
-                obj.update(raw=raw_item)
-                continue
+    def process_websocket_data(self, data: WebsocketData) -> Event | str:
+        """Process and forward websocket data."""
+        if data.meta.message == MessageKey.EVENT:
+            return self.process_event(data.event)
 
-            self._items[key] = self.item_cls(raw_item, self.controller.request)
-            new_items.add(key)
+        if data.meta.message in self.process_messages:
+            return self.process_item(data.data["data"])
 
-            for callback in self._subscribers:
-                callback("added", key)
+        if data.meta.message in self.remove_messages:
+            return self.remove_item(data.data["data"])
 
-        return new_items
+        return ""
```

Comment:
I would have used None as return value here and in the three methods below to indicate nothing is returned.
```suggestion
        return None
```

Reply:
This is just to support current functionality, this will be completely replaced by the subscribe functionality, but I don't want to break things too early on.